### DISCUSSION TITLE
Welcome to J2534-Sharp Discussions!

### DISCUSSION THREAD

**Discussion by BrianHumlicek** [BrianHumlicek/J2534-Sharp#11] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jglim** [comment]
Hi Brian,

Just wanted to thank you for a great library! It is [in use](https://github.com/BrianHumlicek/J2534-Sharp#feedback) in a [small open-source diagnostics tool](https://github.com/jglim/CaesarSuite/), and it's a good feeling to be able to focus on the reverse-engineering aspect of my project instead of the J2534 plumbing :)

**Reply by BrianHumlicek** [reply]
Thanks for letting me know.  I built this exactly for this reason; so you can get right to the meat of your J2534 project without having to learn all the intricacies of the unmanaged API.

You're welcome, and best of luck with your project!

Brian

**Comment by waynemodz** [comment]
Hey Brian, 
Interesting project you got going on, this is related to my space and unfortunately c# is my only language.
I see the package is no longer under the the nuget repo, 

tried to run a test here ,
` 
                byte Msg = 0x0C;
                byte Cks = Checksums.CalculateISO9141(new byte[] { 68, Msg });
                
                Channel.SendMessage(new byte[] { 0x33, Msg, Cks });
                GetMessageResults Response = Channel.GetMessage();
                foreach (var M in Response.Messages)
                {
                    PrintF("RESP: "+M.Data.Message);
                }
`

`        public static byte CalculateISO9141(byte[] data)
        {
            byte checksum = 0;

            foreach (byte b in data)
            {
                checksum ^= b; // XOR operation
            }

            return checksum;
        }`

i get an INDEX ARRAY error, 
and then the device wont connect until i restart the program.
Is there any further documentation or examples?

**Reply by BrianHumlicek** [reply]
What line is the exception happening on and what exactly is the exception?  Are you sure this is a problem with J2534Sharp?

Brian

**Reply by waynemodz** [reply]
thank you for your reply, so this car i'm testing on uses  KWP2000 which should be ISO9141, i'm getting a BUFFER_EMPTY response.  0x0C should be engine speed rpm, so was just doing a test too get that. Only data i managed to get was battery voltage.

**Comment by czechu82** [comment]
Hi Brian,

I'm trying to build simple diagnostic app . 
How do i set up can ID and filters correctly ?
Example :
Tool Id -> 1F 40 10 F2
ECU id  -> 1F 40 F2 10 
ISO15765 , Baud 500 , channel 1 

your code :
 MessageFilter FlowControlFilter = new MessageFilter()
            {
                FilterType = Filter.FLOW_CONTROL_FILTER,
                Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
                Pattern = new byte[] { 0x00, 0x00, 0x07, 0xE8 },
                FlowControl = new byte[] { 0x00, 0x00, 0x07, 0xE0 }
            };

My main  goal is to read Raw data coming from the server. 
How can this be done using your code ?
     GetMessageResults Response = Channel.GetMessage();  ?

Also can you explain bytes in  this command:
Channel.SendMessage(new byte[] { 0x00, 0x00, 0x07, 0xE0, 0x01, 0x00 });
looks like first 4 bytes are ID and last 2  (0x01,0x00) are raw data ?

Regards

**Reply by BrianHumlicek** [reply]
Your Tool and ECU addresses appear more like J1850 addresses than they do CAN/ISO15765.  If they are in fact CAN, they are 29bit.  If so, you need to include that in the connect flag as well as in the TxFlag when sending.

Additionally, the ToolId and ECU Id don't match the flow control filter, which I am guessing is just a copy/paste from my example.

If you want to monitor all bus traffic, you need to set a PASS ALL filter.  This is most easily done by the following or something like it;

MessageFilter passAll = new MessageFilter(UserFilterType.PASSALL, null);
passAll.TxFlags |= TxFlag.CAN_29BIT_ID;
channel.StartMsgFilter(passAll);

Brian

**Reply by BrianHumlicek** [reply]
I forgot to mention, if your goal is to read raw data, you probably want to connect as CAN vs ISO15765, as I don't believe the PASS ALL filter is applicable to ISO15765.

Honestly, it sounds like you need to do some more reading about OBD diagnostics.  I can try to help got you pointed in the right direction, but at the end of the day, the J2534 standard is what you need to know.  My library doesn't change that.

Brian

**Reply by czechu82** [reply]
I did some reading about UDS and it doesnt seem to be complicated . It's pretty straight forward.
I worked with UDS using rp1210 standards and now i want to do same thing using pass through device , so  my only issue here is to set up filters and build correct connection string. Rest is easy .


By saying read all raw data I meant only data from defined ID returned from server after each request.
I can format that data to display only UDS response like 11-01 -> 51-01 
Sorry for not being specific .


Here is what i have tried with error 

MessageFilter FlowControlFilter = new MessageFilter()
{
FilterType = Filter.FLOW_CONTROL_FILTER,
Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF }, 
Pattern = new byte[] { 0x1F, 0x40, 0xF2, 0x10 },
FlowControl = new byte[] { 0x1F, 0x40, 0x10, 0xF2},
TxFlags = TxFlag.CAN_29BIT_ID

};

    string DllFileName = APIFactory.GetAPIinfo().First().Filename;

            using (API API = APIFactory.GetAPI(DllFileName))
            using (Device Device = API.GetDevice())

// not sure about this connection string 
            Using Channel As Channel = Device.GetChannel(Protocol.ISO15765, Baud.ISO15765_500000, ConnectFlag.CAN_29BIT_ID))
            {
                Channel.StartMsgFilter(FlowControlFilter);
                Console.WriteLine($"Voltage is {Channel.MeasureBatteryVoltage() / 1000}");    

// trying to send  reset command    

Channel.SendMessage(new byte[] { 0x1F, 0x40, 0xF2, 0x10, 0x11, 0x01 });
                GetMessageResults Response = Channel.GetMessage();
            }
        }

**Reply by BrianHumlicek** [reply]
You need to set the 29bit flag in txflags on the channel too.

**Reply by czechu82** [reply]
Ok. Got it .

Channel.DefaultTxFlag = TxFlag.CAN_29BIT_ID

But now I'm not getting any response from ECU.

Looks like I'm trying to send a message in wrong format and ecu doesnt recognize it .

For example after sending this command :

Channel.SendMessage(new byte[] { 0x1F, 0x40, 0x10, 0xF2, 0x10, 0x03 });



 command sent using j2534 dll with no response 
17:13:47:0940 Rx 1 0x1F4010F2 x 3 02 10 03 


 command sent with other tool with  positive response
17:13:54:6000 Rx 1 0x1F4010F2 x 8 02 10 03 00 00 00 00 00  
17:13:54:6080 Rx 1 0x1F40F210 x 8 06 50 03 00 19 01 F4 00 

What else am I missing ?

**Reply by czechu82** [reply]
seems like i found the issue .

"Be aware, many engine controllers are sensitive to the format of the flow control message. Some controllers expect the messages to be zero‑padded to a full 8 bytes, while others expect only the necessary data. Adjust the ISO15765_FRAME_PAD flag to match the expected behavior."


here's what i have now 

setting filters:

  ```
    FilterType = Filter.FLOW_CONTROL_FILTER,
            Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
            Pattern = new byte[] { 0x1F, 0x40, 0x10, 0xF2 },
            FlowControl = new byte[] { 0x1F, 0x40, 0xF2, 0x10 },
            TxFlags = TxFlag.CAN_29BIT_ID
```

connection string:

```
using (Channel Channel = Device.GetChannel(Protocol.ISO15765, Baud.ISO15765_500000, ConnectFlag.CAN_29BIT_ID))
  {
            Channel.DefaultTxFlags |= TxFlag.CAN_29BIT_ID;
            Channel.StartMsgFilter(FlowControlFilter);
            richTextBox1.AppendText($"Voltage is {Channel.MeasureBatteryVoltage() / 1000}");
            Channel.SendMessage(new byte[] { 0x1F, 0x40, 0x10, 0xF2, 0x10, 0x03,});
            GetMessageResults Response = Channel.GetMessage();
        }

```

How do i use ISO15765_FRAME_PAD in this code ?

**Reply by BrianHumlicek** [reply]
Probably something like this

```
    MessageFilter FlowControlFilter = new MessageFilter()
    {
        FilterType = Filter.FLOW_CONTROL_FILTER,
        Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
        Pattern = new byte[] { 0x1F, 0x40, 0xF2, 0x10 },
        FlowControl = new byte[] { 0x1F, 0x40, 0x10, 0xF2 },
        TxFlags = TxFlag.CAN_29BIT_ID | TxFlag.ISO15765_FRAME_PAD
    };
```

**Reply by czechu82** [reply]
Thanks Brian !

It works but I had to set it on the channel.

I'm still having small issues filtering messages. 
For some reason empty massages  are added to queue.
 
Based on this article only pattern ID should be added to queue.
https://www.opusivs.com/support/j2534-support/passthru-sae-j2534-dll/passthrustartmsgfilter

        Pattern = new byte[] { 0x1F, 0x40, 0xF2, 0x10 },

I'm trying to read raw data and my current output looks like this :

1F 40 10 F2 10 03 
1F 40 10 F2  <- this goes to queue after 10 03 request and when i try to read UDS response it shows empty message.
1F 40 F2 10 50 03 00 19 01 F4 <- 2nd try shows correct response 

Is it wrong filter set-up or something else?


Regards

**Reply by BrianHumlicek** [reply]
Try adding this before you send the message.  
`
Channel.SetConfig(Parameter.LOOP_BACK, 0);
`
Additionally, you may need to discard received messages with RX flags set.  These are sometimes used as tx receipts by the device.  They aren't physical messages that are received over the bus.

**Reply by czechu82** [reply]
Setting LOOP_BACK, 0 didn't work. I believe its set to 0 by default.
When set to 1 it adds 00 00 00 00 to queue.

Want to set RX flags but cant find any info how its done.

**Reply by czechu82** [reply]
For now i fixed it in the app by setting condition if message length <= 4 then read response again , but still I'd like to fix it the proper way by setting RX flags if possible .

**Reply by BrianHumlicek** [reply]
You don’t set RXflags, you read them.  Instead of checking if the length is greater than 3, check if the flags are clear (equal to zero).
